### PR TITLE
Add file name to DirectAccessUrl

### PR DESCRIPTION
**Pull request by canpan14**
This will cause it to be returned in the response

### CODE REVIEW

**Comment by canpan14** [comment]
@tpage-alfresco @jamalkm Let me know if this is what you were looking for. API Specs would be modified as well of course.

File: repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java [hunk 1]
```diff
@@ -166,34 +166,46 @@ public void testRequestContentDirectUrl_StoreIsEnabledButNotImplemented()
...
+        mockDirectAccessUrl.setAttachment(true);
+        mockDirectAccessUrl.setContentUrl(SOME_CONTENT_URL);
         final String fileNameOverride = "fileNameOverride.txt";
         setupSystemWideDirectAccessConfig(ENABLED);
         when(mockContentStore.isContentDirectUrlEnabled()).thenReturn(ENABLED);
+        when(mockContentStore.requestContentDirectUrl(anyString(), eq(true), eq(SOME_FILE_NAME), anyString(), anyLong())).thenReturn(mockDirectAccessUrl);
 
         DirectAccessUrl directAccessUrl = contentService.requestContentDirectUrl(NODE_REF, PROP_CONTENT_QNAME, true, 20L, fileNameOverride);
-        assertNull(directAccessUrl);
+        assertEquals("fileName was not set properly on the DirectAccessUrl response", fileNameOverride, directAccessUrl.getFileName());
```
Comment (mpichura) [inline repository/src/test/java/org/alfresco/repo/content/ContentServiceImplUnitTest.java:178]: "fileName was not set properly on the DirectAccessUrl response" could be a private constant in this test class as this particular string is repeated 3 times.

**Comment by tpage-alfresco** [comment]
PMD says that the number of issues in these files went from 36 to 33 - you fixed three issues - thanks!
```
--- 3 resolved JUnitAssertionsShouldIncludeMessage violation(s) ---
```

**Review by tpage-alfresco — APPROVED** [review]
LGTM

**Comment by canpan14** [comment]
> Ah - spoke too soon. The existing tests in ContentServiceImplUnitTest.java fail with NPEs:
> 
> ```
> Error:  Tests run: 1175, Failures: 0, Errors: 3, Skipped: 2, Time elapsed: 45.869 s <<< FAILURE! - in org.alfresco.AllUnitTestsSuite
> Error:  testRequestContentDirectUrl_SystemWideIsEnabledButStoreIsDisabled(org.alfresco.repo.content.ContentServiceImplUnitTest)  Time elapsed: 0.004 s  <<< ERROR!
> java.lang.NullPointerException: Cannot invoke "org.alfresco.service.cmr.repository.DirectAccessUrl.setFileName(String)" because "directAccessUrl" is null
> 	at org.alfresco.repo.content.ContentServiceImplUnitTest.testRequestContentDirectUrl_SystemWideIsEnabledButStoreIsDisabled(ContentServiceImplUnitTest.java:150)
> 
> Error:  testRequestContentDirectUrl_StoreRequestContentDirectUrlIsCalledWithFileNameOverride(org.alfresco.repo.content.ContentServiceImplUnitTest)  Time elapsed: 0.002 s  <<< ERROR!
> java.lang.NullPointerException: Cannot invoke "org.alfresco.service.cmr.repository.DirectAccessUrl.setFileName(String)" because "directAccessUrl" is null
> 	at org.alfresco.repo.content.ContentServiceImplUnitTest.testRequestContentDirectUrl_StoreRequestContentDirectUrlIsCalledWithFileNameOverride(ContentServiceImplUnitTest.java:177)
> 
> Error:  testRequestContentDirectUrl_StoreIsEnabledButNotImplemented(org.alfresco.repo.content.ContentServiceImplUnitTest)  Time elapsed: 0.001 s  <<< ERROR!
> java.lang.NullPointerException: Cannot invoke "org.alfresco.service.cmr.repository.DirectAccessUrl.setFileName(String)" because "directAccessUrl" is null
> 	at org.alfresco.repo.content.ContentServiceImplUnitTest.testRequestContentDirectUrl_StoreIsEnabledButNotImplemented(ContentServiceImplUnitTest.java:161)
> ```
> 
> Edit: There's also a NPE in this application context test:
> 
> ```
> Error:  Tests run: 452, Failures: 0, Errors: 1, Skipped: 1, Time elapsed: 345.725 s <<< FAILURE! - in org.alfresco.AppContext04TestSuite
> Error:  testWhenRequestContentDirectUrlIsNotSupported(org.alfresco.repo.version.ContentServiceImplTest)  Time elapsed: 0.74 s  <<< ERROR!
> java.lang.NullPointerException: Cannot invoke "org.alfresco.service.cmr.repository.DirectAccessUrl.setFileName(String)" because "directAccessUrl" is null
> 	at org.alfresco.repo.version.ContentServiceImplTest.testWhenRequestContentDirectUrlIsNotSupported(ContentServiceImplTest.java:185)
> ```

I swear I ran these tests and they passed locally. I'll get them fixed!

**Comment by canpan14** [comment]
Out of curiosity @tpage-alfresco, what's the process for changes getting into the next release (7.5 for example). Is that something I also need to assist with, or is that a purely internal process?

File: data-model/src/main/java/org/alfresco/service/cmr/repository/DirectAccessUrl.java [hunk 2]
```diff
@@ -70,18 +71,28 @@ public void setAttachment(boolean attachment)
...
+    }
+
     @Override public boolean equals(Object obj)
     {
         if (this == obj) return true;
         if (obj == null || getClass() != obj.getClass()) return false;
 
         DirectAccessUrl that = (DirectAccessUrl) obj;
-        return attachment == that.attachment && Objects.equals(contentUrl,
+        return fileName.equals(that.fileName) && attachment == that.attachment && Objects.equals(contentUrl,
```
Comment (pzhyland) [inline data-model/src/main/java/org/alfresco/service/cmr/repository/DirectAccessUrl.java:90]: There is nothing preventing the `fileName` from being `null` so `NullPointerException` is possible here.
```suggestion
        return Objects.equals(fileName, that.fileName) && attachment == that.attachment && Objects.equals(contentUrl,
```

**Comment by tpage-alfresco** [comment]
> Out of curiosity @tpage-alfresco, what's the process for changes getting into the next release (7.5 for example). Is that something I also need to assist with, or is that a purely internal process?

@canpan14 This should be in the next release from master - currently planned as 23.x (to match with the Hyland versioning scheme for other products). I think this will miss the upcoming 7.3.1 service pack, but since this is a new feature then I think that's the correct outcome anyway. There's nothing to do from your side once the code is merged - everything on master is heading towards the 23.x release.

**Comment by tpage-alfresco** [comment]
Two of the NPEs are still there:
```
Error:  Tests run: 1175, Failures: 0, Errors: 2, Skipped: 2, Time elapsed: 54.188 s <<< FAILURE! - in org.alfresco.AllUnitTestsSuite
Error:  testRequestContentDirectUrl_StoreRequestContentDirectUrlIsCalledWithFileNameOverride(org.alfresco.repo.content.ContentServiceImplUnitTest)  Time elapsed: 0.006 s  <<< ERROR!
java.lang.NullPointerException: Cannot invoke "org.alfresco.service.cmr.repository.DirectAccessUrl.setFileName(String)" because "directAccessUrl" is null
	at org.alfresco.repo.content.ContentServiceImplUnitTest.testRequestContentDirectUrl_StoreRequestContentDirectUrlIsCalledWithFileNameOverride(ContentServiceImplUnitTest.java:177)

Error:  testRequestContentDirectUrl_StoreIsEnabledButNotImplemented(org.alfresco.repo.content.ContentServiceImplUnitTest)  Time elapsed: 0.002 s  <<< ERROR!
java.lang.NullPointerException: Cannot invoke "org.alfresco.service.cmr.repository.DirectAccessUrl.setFileName(String)" because "directAccessUrl" is null
	at org.alfresco.repo.content.ContentServiceImplUnitTest.testRequestContentDirectUrl_StoreIsEnabledButNotImplemented(ContentServiceImplUnitTest.java:161)
```

**Comment by canpan14** [comment]
> Two of the NPEs are still there:
> 
> ```
> Error:  Tests run: 1175, Failures: 0, Errors: 2, Skipped: 2, Time elapsed: 54.188 s <<< FAILURE! - in org.alfresco.AllUnitTestsSuite
> Error:  testRequestContentDirectUrl_StoreRequestContentDirectUrlIsCalledWithFileNameOverride(org.alfresco.repo.content.ContentServiceImplUnitTest)  Time elapsed: 0.006 s  <<< ERROR!
> java.lang.NullPointerException: Cannot invoke "org.alfresco.service.cmr.repository.DirectAccessUrl.setFileName(String)" because "directAccessUrl" is null
> 	at org.alfresco.repo.content.ContentServiceImplUnitTest.testRequestContentDirectUrl_StoreRequestContentDirectUrlIsCalledWithFileNameOverride(ContentServiceImplUnitTest.java:177)
> 
> Error:  testRequestContentDirectUrl_StoreIsEnabledButNotImplemented(org.alfresco.repo.content.ContentServiceImplUnitTest)  Time elapsed: 0.002 s  <<< ERROR!
> java.lang.NullPointerException: Cannot invoke "org.alfresco.service.cmr.repository.DirectAccessUrl.setFileName(String)" because "directAccessUrl" is null
> 	at org.alfresco.repo.content.ContentServiceImplUnitTest.testRequestContentDirectUrl_StoreIsEnabledButNotImplemented(ContentServiceImplUnitTest.java:161)
> ```

Had to switch IDEs to get these failures to showup. For some reason the tests passed in VS Code but failed in IntelliJ. The were correct to fail and I've fixed them now, but very odd. I usually don't do java in VS Code to be fair, lesson learned!

Note I wrapped the set file name in a null check because it seemed like one of the preexisting unit tests made it a possibility that `requestContentDirectUrl` could be enabled on a content store, but not implemented. If that's not actually true let me know and I'll re-modify the code to fix that unit test.

**Review by tpage-alfresco — APPROVED** [review]
LGTM

**Comment by canpan14** [comment]
@tpage-alfresco I assume we are waiting for https://github.com/Alfresco/rest-api-explorer/pull/197 to go through at the same time?

**Comment by tpage-alfresco** [comment]
> @tpage-alfresco I assume we are waiting for [Alfresco/rest-api-explorer#197](https://github.com/Alfresco/rest-api-explorer/pull/197) to go through at the same time?

Hi @canpan14. Thanks for your patience on this! We're a bit busy this week with the service pack release of ACS 7.4.1, but I'm hopeful we can get around to the final merge on Monday. At this point some further tests will run against the change (e2e tests mainly focused around compatibility of the Repo with other components), but I'm not expecting any issues there. Thanks again for your contribution!

**Comment by canpan14** [comment]
> > @tpage-alfresco I assume we are waiting for [Alfresco/rest-api-explorer#197](https://github.com/Alfresco/rest-api-explorer/pull/197) to go through at the same time?
> 
> Hi @canpan14. Thanks for your patience on this! We're a bit busy this week with the service pack release of ACS 7.4.1, but I'm hopeful we can get around to the final merge on Monday. At this point some further tests will run against the change (e2e tests mainly focused around compatibility of the Repo with other components), but I'm not expecting any issues there. Thanks again for your contribution!

Thanks again! No rush, just wanted to make there were no further actions expected from me.

**Pull request merged**